Userland/Libraries/LibWeb/Loader/ResourceLoader.cpp

```diff
@@ -250,6 +255,61 @@ void ResourceLoader::load(LoadRequest& request, Function<void(ReadonlyBytes, Has
+                Core::DirIterator dt(lexical_path.string(), Core::DirIterator::Flags::SkipParentAndBaseDir);
+                Vector<DeprecatedString> names;
+                while (dt.has_next())
+                    names.append(dt.next_path());
+                quick_sort(names);
+
+                StringBuilder contents;
+                contents.append("<table>"sv);
+                for (auto& name : names) {
+                    auto path = lexical_path.append(name);
+                    struct stat st;
+                    memset(&st, 0, sizeof(st));
+                    stat(path.string().characters(), &st);
+
+                    contents.append("<tr>"sv);
+                    contents.appendff("<td><span class=\"{}\"></span></td>", S_ISDIR(st.st_mode) ? "folder" : "file");
+                    contents.appendff("<td><a href=\"file://{}\">{}</a></td><td>&nbsp;</td>"sv, path, name);
+                    contents.appendff("<td>{:10}</td><td>&nbsp;</td>", st.st_size);
+                    contents.appendff("<td>{}</td>"sv, Core::DateTime::from_timestamp(st.st_mtime).to_deprecated_string());
+                    contents.append("</tr>\n"sv);
+                }
+                contents.append("</table>"sv);
+
+                // Generate HTML directory page from directory template file
+                // FIXME: Use an actual templating engine (our own one when it's built, preferably with a way to check these usages at compile time)
+                auto template_path = AK::URL::create_with_url_or_path(FrameLoader::directory_page_url()).serialize_path();
+                auto maybe_template_file = Core::File::open(template_path, Core::File::OpenMode::Read);
+                if (maybe_template_file.is_error()) {
+                    log_failure(request, maybe_template_file.error());
+                    if (error_callback)
+                        error_callback(DeprecatedString::formatted("{}", maybe_template_file.error()), 500u);
+                    return;
+                }
+                auto template_file = maybe_template_file.release_value();
+                auto template_contents = template_file->read_until_eof().release_value_but_fixme_should_propagate_errors();
+                StringBuilder builder;
+                SourceGenerator generator { builder };
+                generator.set("path", escape_html_entities(lexical_path.string()));
+                generator.set("parent_path", escape_html_entities(lexical_path.parent().string()));
+                generator.set("contents", contents.to_deprecated_string());
+                generator.append(template_contents);
+
+                // Return HTML response
+                HashMap<DeprecatedString, DeprecatedString, CaseInsensitiveStringTraits> response_headers;
+                response_headers.set("Content-Type"sv, "text/html"sv);
+                success_callback(generator.as_string_view().bytes(), response_headers, {});
+                return;
+            }
```

Comment:
Conceptually, why does this all live in `ResourceLoader`? Generating HTML for a directory seems specific enough that it warrants its own class.

Reply:
You are right, I have split of the response generation code to a new singleton class called `FileDirectoryLoader.cpp`